rfd/0177-upgrade-js-package-manager.md

```diff
@@ -0,0 +1,183 @@
+For example, when the developer installs package `fuu`, which depends on 
+`bar` and `baz`, the package manager creates three directories, 
+`fuu`, `bar` and `baz` in `node_modules`.
+The developer can then import all of these packages in code, although only one 
+of them is listed in the `package.json` file.
+It is possible because of how the Node.js module resolution algorithm works. 
+The `package.json` file is not needed at all; when Node.js resolves an import, 
+it traverses `node_modules` folders all the way to the user's root directory, 
+looking for that dependency.
+
+The main problem with using undeclared dependencies is that we have no control 
+over them. The next (even a patch) version of the `fuu` package may update `bar` 
+to a version that makes our code relying on `bar` stop working properly (read 
+more: [pnpm's strictness helps to avoid silly bugs](https://www.kochan.io/nodejs/pnpms-strictness-helps-to-avoid-silly-bugs.html)).
+We can see such dependencies in our codebase: we import `prop-types` in a few 
+components, but we don't specify in any workspace’s `package.json` 
+(it's taken from `react-select`).
+
+To address that problem, the new package manager should use non-flat (or 
+sometimes called isolated) `node_modules` structure, which contains only 
+direct dependencies, declared in `package.json`.
+Let's go through installation options of the most popular package managers:
+* npm has `install-strategy`:
+    * hoisted (default): install non-duplicated in top-level, and duplicated as 
+  necessary within directory structure.
+    * nested: install in place, no hoisting. Note: this does not seem to work 
+  correctly https://github.com/npm/cli/issues/6537.
+    * shallow: only install direct deps at top-level. Note: it creates 
+  `node_modules` directory only in the project's root, so it does not change 
+  much when it comes to workspaces, the dependencies are still hoisted.
+    * linked: (experimental) install in node_modules/.store, link in place, 
+  unhoisted.
+  Note: workes like pnpm. See an [RFD](https://github.com/npm/rfcs/blob/main/accepted/0042-isolated-mode.md).
+* Yarn has `nodeLinker`: 
+    * pnp: no node_modules, unhoisted. Note: does not work with Electron.
+    * pnpm: install in `node_modules/.store`, link in place, unhoisted. 
+  Note: It should work with the latest alpha electron-builder and Electron 
+  versions.
+    * node_modules: flat, hoisted. Same as the node_modules created by npm or 
+  Yarn Classic.
+* pnpm has `node-linker`:
+    * isolated: dependencies are symlinked from a virtual store at
+  `node_modules/.pnpm`. Note: I was able to build Teleport Connect with the 
+  latest electron-builder and Electron versions.
+    * hoisted: a flat node_modules without symlinks is created. 
+  Same as the node_modules created by npm or Yarn Classic.
+    * pnp: no node_modules. Same as pnp in Yarn.
+
+The installation option that is most strict and works with Electron is 
+`linked`/`pnpm`. 
```

Comment:
You mean npm's `linked` and Yarn's `pnpm`?

Reply:
I meant npm's `linked`, Yarn's `pnpm` and pnpm's `isolated`. I will fix that.